docs/design-documents/20221027-builtin-config-validation.md

```diff
@@ -0,0 +1,144 @@
+        TYPE_LESS_THAN = 3;
+        // Parameter must be included in the comma separated list {value}.
+        TYPE_INCLUSION = 4;
+        // Parameter must not be included in the comma separated list {value}.
+        TYPE_EXCLUSION = 5;
+        // Parameter must match the regex {value}.
+        TYPE_REGEX = 6;
+      }
+
+      Type type = 1;
+      // The value to be compared with the parameter,
+      // or a comma separated list in case of Validation.TYPE_INCLUSION or Validation.TYPE_EXCLUSION.
+      string value = 2;
+    }
+
+    // Type shows the parameter type.
+    enum Type {
+      TYPE_UNSPECIFIED = 0;
+      // Parameter is a string.
+      TYPE_STRING = 1;
+      // Parameter is a number.
+      TYPE_NUMBER = 2;
+      // Parameter is a boolean.
+      TYPE_BOOL = 3;
+      // Parameter is a file.
+      TYPE_FILE = 4;
+      // Parameter is a duration.
+      TYPE_DURATION = 5;
+    }
+
+    string description = 1;
+    string default = 2;
+    Type type = 3;
+    repeated Validation validations = 4;
+  }
+
+  string name = 1;
+  string summary = 2;
+  string description = 3;
+  string version = 4;
+  string author = 5;
+  map<string, Parameter> destination_params = 6;
+  map<string, Parameter> source_params = 7;
+}
+```
+
+For now, this is only used when listing the plugins in Conduit. All these validation options are not yet implemented and
+are not exposed by the SDK.
+
+## Scope
+
+Implementing the validation options provided by the proto file. So, giving the developer the option to specify
+validations for each parameter and Conduit will make sure to run the validations.
+
+Validating the type of the parameter, we have 5 types supported in the proto design {string, number, bool, file, and duration}
+
+Providing a utility function to generate the config map for the `Parameters` function from a config struct. This is not 
+mandatory for the scope of this feature, but it would be a nice to have and would make the developing experience for connectors easier.
+
+## Questions:
+
+**Q**: Should we implement the validations from the SDK or Conduit side?
+
+**A**: If we implement it from Conduit side, this means we can add more validations in the future without changing the SDK.
+However, implementing it from the SDK side means that all the validations will happen on the connector side, using the 
+Config function, and the developer will still have the ability to add custom validations for some specific cases that 
+the SDK wouldn't cover. So we decided on implementing validating from the SDK side.
+
+**Q**: Should generating the Configurations from a Go struct be in the scope of this feature?
+
+**A**: We decided to add the generation as a subtask for now, and see how much time left we have for it. We agreed that this
+would be super helpful for developers and easier to use, and will continue adding to the feature in the future.
+
+## Implementation
+
+Implementing this feature consists of three main steps:
```

Comment:
I think there's another step (well at least I'd mark it as a separate step) - we need to make sure the connector protocol is adjusted and that the validations for parameters are added there. I'd also explicitly state that the change needs to be done in a backwards compatible way.

Reply:
I didn't go into much details to explicitly mention the connector protocol, the only change I see would be adding the `Validation` as a struct, and then adding `validations` slice into the parameter. did I miss something?
and agreed, will add this as a separate step!

Reply:
Well, it will affect the connector protocol, the SDK as well as Conduit. It's a decent amount of work, worth to make it a separate step IMO 🙂 Also the backwards compatibility part will mean we need to parse the old `required` field and convert it into a validation, that can be part of this. It's probably the first thing that will need to happen so we have the validation structs we need in the SDK to populate the parameter specs.

Reply:
@lovromazgon updated